src/ray/raylet/node_manager.cc

```diff
@@ -1406,7 +1406,15 @@ void NodeManager::ProcessAnnounceWorkerPortMessage(
 
 void NodeManager::HandleWorkerAvailable(const std::shared_ptr<ClientConnection> &client) {
   std::shared_ptr<WorkerInterface> worker = worker_pool_.GetRegisteredWorker(client);
-  HandleWorkerAvailable(worker);
+  if (worker) {
+    HandleWorkerAvailable(worker);
+  } else {
+    RAY_LOG(INFO)
+        << "[" << client->GetDebugLabel()
+        << "]NodeManager::HandleWorkerAvailable: Received worker available message but "
+           "no associated worker found. This worker process may leak. The worker process "
+           "has either not been registered yet or the process has already exited.";
```

Comment:
What's the sequence of events that can make this happen?

Reply:
I don't know but I think it's too complicated to figure it out. This seems to me to be the safer option.

Reply:
I did a quick check of all callers of `HandleWorkerAvailable()`. I think the only case where this check might fail is in `ProcessClientMessage`:

```
case protocol::MessageType::TaskDone: {
    HandleWorkerAvailable(client);
  } break;
```

Can we move the worker is null if else there and still keep the check inside `HandleWorkerAvailable()`?


I think it's enough for the release blocker but we should still keep the ticket open until we figure out exactly how this can happen.

Reply:
> I did a quick check of all callers of HandleWorkerAvailable(). I think the only case where this check might fail is in ProcessClientMessage:

I don't think that is correct. That call to HandleWorkerAvailable() enters this HandleWorkerAvailable(ClientConnection), which tries to find the Worker associated with the ClientConnection. We don't find a worker so then HandleWorkerAvailable(Worker) fails on the assertion.

So let's keep it this way?

> I think it's enough for the release blocker but we should still keep the ticket open until we figure out exactly how this can happen.

We can keep the ticket open, but we don't have a reproduction and the worker pool code is difficult to understand. I doubt it will be very easy to find the root cause.

Reply:
I think this sounds good. https://github.com/ray-project/ray/pull/41841#discussion_r1425797171. Let's just do it and merge it then?